pkg/listeners/malformed_host_override.go

```diff
@@ -0,0 +1,85 @@
+// +build !windows
+
+package listeners
+
+import (
+	"net"
+	"strings"
+)
+
+// MalformedHostHeaderOverride is a wrapper to be able
+// to overcome the 400 Bad request coming from old docker
+// clients that send an invalid Host header.
+type MalformedHostHeaderOverride struct {
+	net.Listener
+}
+
+// MalformedHostHeaderOverrideConn wraps the underlying unix
+// connection and keeps track of the first read from http.Server
+// which just reads the headers.
+type MalformedHostHeaderOverrideConn struct {
+	net.Conn
+	first bool
+}
+
+// Read reads the first *read* request from http.Server to inspect
+// the Host header. If the Host starts with / then we're talking to
+// an old docker client which send an invalid Host header. To not
+// error out in http.Server we rewrite the first bytes of the request
+// to sanitize the Host header itself.
+// In case we're not dealing with old docker clients the data is just passed
+// to the server w/o modification.
+func (l *MalformedHostHeaderOverrideConn) Read(b []byte) (n int, err error) {
+	// http.Server uses a 4k buffer
+	if l.first && len(b) == 4096 {
+		// This keeps track of the first read from http.Server which just reads
+		// the headers
+		l.first = false
+		// The first read of the connection by http.Server is done limited to
+		// DefaultMaxHeaderBytes (usually 1 << 20) + 4096.
+		// Here we do the first read which gets us all the http headers to
+		// be inspected and modified below.
+		c, err := l.Conn.Read(b)
+		if err != nil {
+			return c, err
+		}
+		parts := strings.Split(string(b[:c]), "\n")
+		head := []string{parts[0]}
+		if len(parts) > 0 {
+			if !strings.HasPrefix(parts[1], "Host:") {
+				// old docker clients sends the Host header at parts[1]
+				// which is the second line of the http request
+				// if we're not talking to an old docker client, just skip
+				head = parts
+			} else if !strings.HasPrefix(parts[1], "Host: /") {
+				// we're talking to a newer docker clients if Host doesn't start
+				// with a slash
+				head = parts
+			} else {
```

Comment:
They both have the same logic. What case am I missing?

| parts[1]    | current code | new code     |
|-------------|--------------|--------------|
| foo         | head=parts   | head=parts   |
| Host: bar | head=parts   | head=parts   |
| Host: /baz  | handle stuff | handle stuff |